File: internal/core/dedup.go
package core

import (
	"path/filepath"
	"regexp"
	"strings"
)

type DedupGroup struct {
	Representative FileRecord   `json:"representative"`
	Duplicates     []FileRecord `json:"duplicates"`
	Reason         string       `json:"reason"`
}

type DedupDetector struct {
	PreferUncompressed bool
	FuzzyThreshold     int
}

func NewDedupDetector(preferUncompressed bool, fuzzyThreshold int) *DedupDetector {
	if fuzzyThreshold <= 0 {
		fuzzyThreshold = 2
	}
	return &DedupDetector{
		PreferUncompressed: preferUncompressed,
		FuzzyThreshold:     fuzzyThreshold,
	}
}

func (d *DedupDetector) Detect(records []FileRecord) []DedupGroup {
	var groups []DedupGroup
	used := make(map[int]bool)

	groups = append(groups, d.exactHashMatch(records, used)...)
	groups = append(groups, d.versionGroupMatch(records, used)...)
	groups = append(groups, d.redundantArchiveMatch(records, used)...)
	groups = append(groups, d.sizeMatch(records, used)...)
	groups = append(groups, d.fuzzyNameMatch(records, used)...)

	return groups
}

func (d *DedupDetector) exactHashMatch(records []FileRecord, used map[int]bool) []DedupGroup {
	byHash := make(map[string][]int)
	for i, r := range records {
		if used[i] {
			continue
		}
		byHash[r.FileHash] = append(byHash[r.FileHash], i)
	}

	var groups []DedupGroup
	for _, indices := range byHash {
		if len(indices) < 2 {
			continue
		}

		repIdx := d.selectRepresentative(records, indices)
		var duplicates []FileRecord
		for _, idx := range indices {
			used[idx] = true
			if idx != repIdx {
				duplicates = append(duplicates, records[idx])
			}
		}

		if len(duplicates) > 0 {
			groups = append(groups, DedupGroup{
				Representative: records[repIdx],
				Duplicates:     duplicates,
				Reason:         "hash_match",
			})
		}
	}
	return groups
}

func (d *DedupDetector) sizeMatch(records []FileRecord, used map[int]bool) []DedupGroup {
	bySize := make(map[int64][]int)
	for i, r := range records {
		if used[i] {
			continue
		}
		bySize[r.FileSize] = append(bySize[r.FileSize], i)
	}

	var groups []DedupGroup
	for _, indices := range bySize {
		if len(indices) < 2 {
			continue
		}

		related := d.findFuzzyRelated(records, indices)
		for _, group := range related {
			if len(group) < 2 {
				continue
			}
			repIdx := d.selectRepresentative(records, group)
			var duplicates []FileRecord
			for _, idx := range group {
				used[idx] = true
				if idx != repIdx {
					duplicates = append(duplicates, records[idx])
				}
			}
			if len(duplicates) > 0 {
				groups = append(groups, DedupGroup{
					Representative: records[repIdx],
					Duplicates:     duplicates,
					Reason:         "size_only",
				})
			}
		}
	}
	return groups
}

func (d *DedupDetector) findFuzzyRelated(records []FileRecord, indices []int) [][]int {
	matched := make(map[int]bool)
	var groups [][]int

	for i, idxA := range indices {
		if matched[idxA] {
			continue
		}
		var group []int
		group = append(group, idxA)
		matched[idxA] = true

		for j := i + 1; j < len(indices); j++ {
			idxB := indices[j]
			if matched[idxB] {
				continue
			}
			nameA := normalizeName(records[idxA].Name)
			nameB := normalizeName(records[idxB].Name)
			if levenshtein(nameA, nameB) <= d.FuzzyThreshold {
				group = append(group, idxB)
				matched[idxB] = true
			}
		}

		groups = append(groups, group)
	}

	return groups
}

func (d *DedupDetector) fuzzyNameMatch(records []FileRecord, used map[int]bool) []DedupGroup {
	var unused []int
	for i := range records {
		if !used[i] {
			unused = append(unused, i)
		}
	}

	var groups []DedupGroup
	matched := make(map[int]bool)

	for i, idxA := range unused {
		if matched[idxA] {
			continue
		}
		var group []int
		group = append(group, idxA)

		for j := i + 1; j < len(unused); j++ {
			idxB := unused[j]
			if matched[idxB] {
				continue
			}
			nameA := normalizeName(records[idxA].Name)
			nameB := normalizeName(records[idxB].Name)
			if levenshtein(nameA, nameB) <= d.FuzzyThreshold {
				group = append(group, idxB)
			}
		}

		if len(group) >= 2 {
			repIdx := d.selectRepresentative(records, group)
			var duplicates []FileRecord
			for _, idx := range group {
				matched[idx] = true
				used[idx] = true
				if idx != repIdx {
					duplicates = append(duplicates, records[idx])
				}
			}
			if len(duplicates) > 0 {
				groups = append(groups, DedupGroup{
					Representative: records[repIdx],
					Duplicates:     duplicates,
					Reason:         "fuzzy_name",
				})
			}
		}
	}

	return groups
}

func (d *DedupDetector) selectRepresentative(records []FileRecord, indices []int) int {
	best := indices[0]

	for _, idx := range indices[1:] {
		if d.isBetterRepresentative(records[idx], records[best]) {
			best = idx
		}
	}

	return best
}

func (d *DedupDetector) isBetterRepresentative(a, b FileRecord) bool {
	if d.PreferUncompressed {
		aUncompressed := isUncompressed(a.Extension)
		bUncompressed := isUncompressed(b.Extension)
		if aUncompressed && !bUncompressed {
			return true
		}
		if !aUncompressed && bUncompressed {
			return false
		}
	}

	aVer, aOk := ExtractVersion(a.Name)
	bVer, bOk := ExtractVersion(b.Name)
	if aOk && bOk {
		cmp := CompareVersions(aVer, bVer)
		if cmp > 0 {
			return true
		}
		if cmp < 0 {
			return false
		}
	}
	if aOk && !bOk {
		return true
	}
	if !aOk && bOk {
		return false
	}

	return a.ModTime.After(b.ModTime)
}

func isUncompressed(ext string) bool {
	uncompressed := map[string]bool{
		".exe": true, ".msi": true, ".pkg": true,
		".dmg": true, ".deb": true, ".rpm": true,
		".jar": true, ".AppImage": true,
	}
	return uncompressed[strings.ToLower(ext)]
}

// versionGroupMatch groups files that share the same base name but differ in version.
// e.g. "Yakit-1.4.6-0417-windows-amd64.exe" and "Yakit-1.4.7-0424-windows-amd64.exe"
func (d *DedupDetector) versionGroupMatch(records []FileRecord, used map[int]bool) []DedupGroup {
	type keyed struct {
		idx  int
		base string
		ver  string
	}
	var versioned []keyed
	for i, r := range records {
		if used[i] {
			continue
		}
		ver, ok := ExtractVersion(r.Name)
		if !ok || ver == "" {
			continue
		}
		base := extractBaseName(r.Name)
		if base == "" {
			continue
		}
		versioned = append(versioned, keyed{idx: i, base: base, ver: ver})
	}

	byBase := make(map[string][]keyed)
	for _, v := range versioned {
		key := v.base + filepath.Ext(records[v.idx].Name)
		byBase[key] = append(byBase[key], v)
	}

	var groups []DedupGroup
	for _, items := range byBase {
		if len(items) < 2 {
			continue
		}
		indices := make([]int, len(items))
		for i, v := range items {
			indices[i] = v.idx
		}
		repIdx := d.selectRepresentative(records, indices)
		var duplicates []FileRecord
		for _, idx := range indices {
			used[idx] = true
			if idx != repIdx {
				duplicates = append(duplicates, records[idx])
			}
		}
		if len(duplicates) > 0 {
			groups = append(groups, DedupGroup{
				Representative: records[repIdx],
				Duplicates:     duplicates,
				Reason:         "multi_version",
			})
		}
	}
	return groups
}

// extractBaseName strips version, platform, arch, and date suffixes from a filename
// to produce a canonical software base name.
// "Yakit-1.4.6-0417-windows-amd64.exe" -> "yakit"
func extractBaseName(filename string) string {
	name := stripExtension(filename)
	// Remove everything from the first version-like segment onward
	re := regexp.MustCompile(`[-_\s]v?\d[\d.]*`)
	loc := re.FindStringIndex(name)
	if loc != nil {
		name = name[:loc[0]]
	}
	name = strings.ToLower(strings.TrimSpace(name))
	return name
}

func normalizeName(name string) string {
	base := name
	ext := filepath.Ext(name)
	if ext != "" {
		base = name[:len(name)-len(ext)]
	}
	base = strings.ToLower(base)
	for _, sep := range []string{"-", "_", ".", " "} {
		base = strings.ReplaceAll(base, sep, "")
	}
	for _, suffix := range []string{"setup", "install", "installer", "win64", "win32", "amd64", "x64", "x86", "64bit", "32bit"} {
		base = strings.TrimSuffix(base, suffix)
	}
	return base
}

func levenshtein(a, b string) int {
	la, lb := len(a), len(b)
	if la == 0 {
		return lb
	}
	if lb == 0 {
		return la
	}

	prev := make([]int, lb+1)
	curr := make([]int, lb+1)

	for j := 0; j <= lb; j++ {
		prev[j] = j
	}

	for i := 1; i <= la; i++ {
		curr[0] = i
		for j := 1; j <= lb; j++ {
			cost := 1
			if a[i-1] == b[j-1] {
				cost = 0
			}
			curr[j] = min3(
				prev[j]+1,
				curr[j-1]+1,
				prev[j-1]+cost,
			)
		}
		prev, curr = curr, prev
	}

	return prev[lb]
}

func min3(a, b, c int) int {
	if a < b {
		if a < c {
			return a
		}
		return c
	}
	if b < c {
		return b
	}
	return c
}

// redundantArchiveMatch detects archive files (.zip, .7z, .rar, etc.) that are
// redundant because an unpackaged version (.exe, .msi, etc.) of the same software exists.
// e.g. "fabric-windows-amd64.exe" vs "fabric.7z"
func (d *DedupDetector) redundantArchiveMatch(records []FileRecord, used map[int]bool) []DedupGroup {
	archiveExts := map[string]bool{
		".zip": true, ".7z": true, ".rar": true, ".gz": true,
		".tar": true, ".xz": true, ".bz2": true,
	}

	type entry struct {
		idx       int
		normalized string
	}

	var archives []entry
	var nonArchives []entry

	for i, r := range records {
		if used[i] {
			continue
		}
		ext := strings.ToLower(r.Extension)
		nameLower := strings.ToLower(r.Name)
		isArchive := archiveExts[ext] ||
			strings.HasSuffix(nameLower, ".tar.gz") ||
			strings.HasSuffix(nameLower, ".tar.xz") ||
			strings.HasSuffix(nameLower, ".tar.bz2")

		norm := normalizeForArchiveMatch(r.Name)
		if norm == "" {
			continue
		}

		if isArchive {
			archives = append(archives, entry{idx: i, normalized: norm})
		} else {
			nonArchives = append(nonArchives, entry{idx: i, normalized: norm})
		}
	}

	var groups []DedupGroup
	usedArchive := make(map[int]bool)

	for _, arch := range archives {
		if usedArchive[arch.idx] {
			continue
		}
		for _, nonArch := range nonArchives {
			if arch.normalized == nonArch.normalized {
				repIdx := nonArch.idx // prefer the non-archive as representative
				dupIdx := arch.idx
				if d.isBetterRepresentative(records[repIdx], records[dupIdx]) {
					// keep non-archive as rep
				} else {
					repIdx, dupIdx = dupIdx, repIdx
				}
				used[repIdx] = true
				used[dupIdx] = true
				usedArchive[arch.idx] = true
				groups = append(groups, DedupGroup{
					Representative: records[repIdx],
					Duplicates:     []FileRecord{records[dupIdx]},
					Reason:         "redundant_archive",
				})
				break
			}
		}
	}
	return groups
}

// normalizeForArchiveMatch produces a canonical name for comparing
// archive vs non-archive versions of the same software.
func normalizeForArchiveMatch(name string) string {
	base := name
	ext := filepath.Ext(name)
	if ext != "" {
		base = name[:len(name)-len(ext)]
	}
	// Strip .tar from double extensions like .tar.gz
	if strings.HasSuffix(strings.ToLower(base), ".tar") {
		base = base[:len(base)-4]
	}
	base = strings.ToLower(base)
	for _, sep := range []string{"-", "_", ".", " "} {
		base = strings.ReplaceAll(base, sep, "")
	}
	for _, suffix := range []string{
		"setup", "install", "installer", "update",
		"win64", "win32", "windowsamd64", "windowsx64", "windowsx86",
		"amd64", "x64", "x86", "64bit", "32bit",
		"linuxamd64", "linuxx64", "darwinamd64", "darwinx64",
	} {
		base = strings.TrimSuffix(base, suffix)
	}
	return base
}

File: src-tauri/src/core/dedup.rs
use crate::core::models::{DedupGroup, FileRecord};
use crate::core::version::{compare_versions, extract_version, levenshtein_distance};
use regex::Regex;
use std::cmp::Ordering;
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::sync::LazyLock;

static VERSION_STRIP_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"[-_\s]v?\d[\d.]*").unwrap());

pub struct DedupDetector {
    pub prefer_uncompressed: bool,
    pub fuzzy_threshold: usize,
}

impl DedupDetector {
    pub fn new(prefer_uncompressed: bool, fuzzy_threshold: usize) -> Self {
        Self {
            prefer_uncompressed,
            fuzzy_threshold: if fuzzy_threshold > 0 { fuzzy_threshold } else { 2 },
        }
    }

    pub fn detect(&self, records: &[FileRecord]) -> Vec<DedupGroup> {
        let mut used: HashSet<usize> = HashSet::new();
        let mut groups = Vec::new();

        groups.extend(self.exact_hash_match(records, &mut used));
        groups.extend(self.version_group_match(records, &mut used));
        groups.extend(self.redundant_archive_match(records, &mut used));
        groups.extend(self.size_match(records, &mut used));
        groups.extend(self.fuzzy_name_match(records, &mut used));

        groups
    }

    fn exact_hash_match(&self, records: &[FileRecord], used: &mut HashSet<usize>) -> Vec<DedupGroup> {
        let mut by_hash: HashMap<String, Vec<usize>> = HashMap::new();
        for (i, r) in records.iter().enumerate() {
            if used.contains(&i) || r.is_app_dir {
                continue;
            }
            by_hash.entry(r.file_hash.clone()).or_default().push(i);
        }

        let mut groups = Vec::new();
        for indices in by_hash.values() {
            if indices.len() < 2 {
                continue;
            }
            let rep_idx = self.select_representative(records, indices);
            let mut duplicates = Vec::new();
            for &idx in indices {
                used.insert(idx);
                if idx != rep_idx {
                    duplicates.push(records[idx].clone());
                }
            }
            if !duplicates.is_empty() {
                groups.push(DedupGroup {
                    representative: records[rep_idx].clone(),
                    duplicates,
                    reason: "hash_match".to_string(),
                });
            }
        }
        groups
    }

    fn version_group_match(&self, records: &[FileRecord], used: &mut HashSet<usize>) -> Vec<DedupGroup> {
        struct Keyed {
            idx: usize,
            base: String,
            ver: String,
        }
        let mut versioned = Vec::new();
        for (i, r) in records.iter().enumerate() {
            if used.contains(&i) {
                continue;
            }
            let (ver, ok) = extract_version(&r.name);
            if !ok || ver.is_empty() {
                continue;
            }
            let base = extract_base_name(&r.name);
            if base.is_empty() {
                continue;
            }
            versioned.push(Keyed { idx: i, base, ver });
        }

        let mut by_base: HashMap<String, Vec<&Keyed>> = HashMap::new();
        for v in &versioned {
            let ext = if let Some(e) = Path::new(&records[v.idx].name).extension() {
                format!(".{}", e.to_string_lossy())
            } else {
                String::new()
            };
            let key = format!("{}{}", v.base, ext);
            by_base.entry(key).or_default().push(v);
        }

        let mut groups = Vec::new();
        for items in by_base.values() {
            if items.len() < 2 {
                continue;
            }
            let indices: Vec<usize> = items.iter().map(|v| v.idx).collect();
            let rep_idx = self.select_representative(records, &indices);
            let mut duplicates = Vec::new();
            for &idx in &indices {
                used.insert(idx);
                if idx != rep_idx {
                    duplicates.push(records[idx].clone());
                }
            }
            if !duplicates.is_empty() {
                groups.push(DedupGroup {
                    representative: records[rep_idx].clone(),
                    duplicates,
                    reason: "multi_version".to_string(),
                });
            }
        }
        groups
    }

    fn redundant_archive_match(&self, records: &[FileRecord], used: &mut HashSet<usize>) -> Vec<DedupGroup> {
        let archive_exts: HashSet<&str> =
            [".zip", ".7z", ".rar", ".gz", ".tar", ".xz", ".bz2"].iter().copied().collect();

        struct Entry {
            idx: usize,
            normalized: String,
        }

        let mut archives = Vec::new();
        let mut non_archives = Vec::new();

        for (i, r) in records.iter().enumerate() {
            if used.contains(&i) {
                continue;
            }
            let ext_lower = r.extension.to_lowercase();
            let name_lower = r.name.to_lowercase();
            let is_archive = archive_exts.contains(ext_lower.as_str())
                || name_lower.ends_with(".tar.gz")
                || name_lower.ends_with(".tar.xz")
                || name_lower.ends_with(".tar.bz2");

            let norm = normalize_for_archive_match(&r.name);
            if norm.is_empty() {
                continue;
            }

            if is_archive {
                archives.push(Entry { idx: i, normalized: norm });
            } else {
                non_archives.push(Entry { idx: i, normalized: norm });
            }
        }

        let mut groups = Vec::new();
        let mut used_archive: HashSet<usize> = HashSet::new();

        for arch in &archives {
            if used_archive.contains(&arch.idx) {
                continue;
            }
            for non_arch in &non_archives {
                if arch.normalized == non_arch.normalized {
                    let rep_idx = non_arch.idx;
                    let dup_idx = arch.idx;
                    used.insert(dup_idx);
                    used_archive.insert(arch.idx);
                    groups.push(DedupGroup {
                        representative: records[rep_idx].clone(),
                        duplicates: vec![records[dup_idx].clone()],
                        reason: "redundant_archive".to_string(),
                    });
                    break;
                }
            }
        }
        groups
    }

    fn size_match(&self, records: &[FileRecord], used: &mut HashSet<usize>) -> Vec<DedupGroup> {
        let mut by_size: HashMap<i64, Vec<usize>> = HashMap::new();
        for (i, r) in records.iter().enumerate() {
            if used.contains(&i) {
                continue;
            }
            by_size.entry(r.file_size).or_default().push(i);
        }

        let mut groups = Vec::new();
        for indices in by_size.values() {
            if indices.len() < 2 {
                continue;
            }
            let related = self.find_fuzzy_related(records, indices);
            for group in related {
                if group.len() < 2 {
                    continue;
                }
                let rep_idx = self.select_representative(records, &group);
                let mut duplicates = Vec::new();
                for &idx in &group {
                    used.insert(idx);
                    if idx != rep_idx {
                        duplicates.push(records[idx].clone());
                    }
                }
                if !duplicates.is_empty() {
                    groups.push(DedupGroup {
                        representative: records[rep_idx].clone(),
                        duplicates,
                        reason: "size_only".to_string(),
                    });
                }
            }
        }
        groups
    }

    fn find_fuzzy_related(&self, records: &[FileRecord], indices: &[usize]) -> Vec<Vec<usize>> {
        let mut matched: HashSet<usize> = HashSet::new();
        let mut groups = Vec::new();

        for (i, &idx_a) in indices.iter().enumerate() {
            if matched.contains(&idx_a) {
                continue;
            }
            let mut group = vec![idx_a];
            matched.insert(idx_a);

            for &idx_b in &indices[i + 1..] {
                if matched.contains(&idx_b) {
                    continue;
                }
                let name_a = normalize_name(&records[idx_a].name);
                let name_b = normalize_name(&records[idx_b].name);
                if levenshtein_distance(&name_a, &name_b) <= self.fuzzy_threshold {
                    group.push(idx_b);
                    matched.insert(idx_b);
                }
            }
            groups.push(group);
        }
        groups
    }

    fn fuzzy_name_match(&self, records: &[FileRecord], used: &mut HashSet<usize>) -> Vec<DedupGroup> {
        let unused: Vec<usize> = (0..records.len()).filter(|i| !used.contains(i)).collect();
        let mut groups = Vec::new();
        let mut matched: HashSet<usize> = HashSet::new();

        for (i, &idx_a) in unused.iter().enumerate() {
            if matched.contains(&idx_a) {
                continue;
            }
            let mut group = vec![idx_a];

            for &idx_b in &unused[i + 1..] {
                if matched.contains(&idx_b) {
                    continue;
                }
                let name_a = normalize_name(&records[idx_a].name);
                let name_b = normalize_name(&records[idx_b].name);
                if levenshtein_distance(&name_a, &name_b) <= self.fuzzy_threshold {
                    group.push(idx_b);
                }
            }

            if group.len() >= 2 {
                let rep_idx = self.select_representative(records, &group);
                let mut duplicates = Vec::new();
                for &idx in &group {
                    matched.insert(idx);
                    used.insert(idx);
                    if idx != rep_idx {
                        duplicates.push(records[idx].clone());
                    }
                }
                if !duplicates.is_empty() {
                    groups.push(DedupGroup {
                        representative: records[rep_idx].clone(),
                        duplicates,
                        reason: "fuzzy_name".to_string(),
                    });
                }
            }
        }
        groups
    }

    fn select_representative(&self, records: &[FileRecord], indices: &[usize]) -> usize {
        let mut best = indices[0];
        for &idx in &indices[1..] {
            if self.is_better_representative(&records[idx], &records[best]) {
                best = idx;
            }
        }
        best
    }

    fn is_better_representative(&self, a: &FileRecord, b: &FileRecord) -> bool {
        if self.prefer_uncompressed {
            let a_unc = is_uncompressed(&a.extension);
            let b_unc = is_uncompressed(&b.extension);
            if a_unc && !b_unc {
                return true;
            }
            if !a_unc && b_unc {
                return false;
            }
        }

        let (a_ver, a_ok) = extract_version(&a.name);
        let (b_ver, b_ok) = extract_version(&b.name);
        if a_ok && b_ok {
            match compare_versions(&a_ver, &b_ver) {
                Ordering::Greater => return true,
                Ordering::Less => return false,
                _ => {}
            }
        }
        if a_ok && !b_ok {
            return true;
        }
        if !a_ok && b_ok {
            return false;
        }

        a.mod_time > b.mod_time
    }
}

fn is_uncompressed(ext: &str) -> bool {
    matches!(
        ext.to_lowercase().as_str(),
        ".exe" | ".msi" | ".pkg" | ".dmg" | ".deb" | ".rpm" | ".jar" | ".appimage"
    )
}

/// Extract base name by stripping version, platform, arch, and date suffixes
pub fn extract_base_name(filename: &str) -> String {
    let name = crate::core::version::strip_extension(filename);
    if let Some(loc) = VERSION_STRIP_RE.find(&name) {
        name[..loc.start()].to_string()
    } else {
        name
    }
    .trim()
    .to_lowercase()
}

pub fn normalize_name(name: &str) -> String {
    let mut base = name.to_string();
    let ext = Path::new(name)
        .extension()
        .map(|e| format!(".{}", e.to_string_lossy()))
        .unwrap_or_default();
    if !ext.is_empty() {
        base = base[..base.len() - ext.len()].to_string();
    }
    base = base.to_lowercase();
    for sep in &["-", "_", ".", " "] {
        base = base.replace(sep, "");
    }
    for suffix in &[
        "setup", "install", "installer", "win64", "win32", "amd64", "x64", "x86", "64bit", "32bit",
    ] {
        if base.ends_with(suffix) {
            base = base[..base.len() - suffix.len()].to_string();
        }
    }
    base
}

fn normalize_for_archive_match(name: &str) -> String {
    let mut base = name.to_string();
    let ext = Path::new(name)
        .extension()
        .map(|e| format!(".{}", e.to_string_lossy()))
        .unwrap_or_default();
    if !ext.is_empty() {
        base = base[..base.len() - ext.len()].to_string();
    }
    let base_lower = base.to_lowercase();
    if base_lower.ends_with(".tar") {
        base = base[..base.len() - 4].to_string();
    }
    let mut base = base.to_lowercase();
    for sep in &["-", "_", ".", " "] {
        base = base.replace(sep, "");
    }
    for suffix in &[
        "setup", "install", "installer", "update",
        "win64", "win32", "windowsamd64", "windowsx64", "windowsx86",
        "amd64", "x64", "x86", "64bit", "32bit",
        "linuxamd64", "linuxx64", "darwinamd64", "darwinx64",
    ] {
        if base.ends_with(suffix) {
            base = base[..base.len() - suffix.len()].to_string();
        }
    }
    base
}
